### PR TITLE
Allow appeals to be searched by their stream docket number

### LINKED ISSUE #15535: Allow appeals to be searched by their stream docket number
## User or job story
As a caseflow user, I need to be able to search for cases by stream docket number, so that I can see all remanded and post decisional appeal streams associated with an appeal.

Currently, new appeal streams share their docket numbers with the original appeal, stored as the appeal's `stream_docket_number`. So even though there are cases with the same "docket number" as the original, only the original is shown.

Search Results | All cases
---|---
Screen Shot 2020-10-30 at 10 41 27 AM|Screen Shot 2020-10-30 at 10 42 25 AM

## Acceptance criteria
- [ ] Please put this work behind the feature toggle: NONE
- [ ] This feature should be accessible to the following user groups: ALL
- [ ] Include screenshot(s) in the Github issue if there are front-end changes
- [ ] Update documentation: 
- [ ] As a user, I can find all appeal streams by searching for the original appeal's docket number

### PR DESCRIPTION
**Pull request by n1zyy**
Resolves #15535

### Description
`AppealFinder.find_appeals_by_docket_number` now also searches for appeals with a matching `stream_docket_number`.

### Acceptance Criteria
(I copied these verbatim. I think we only care about the last one.)

- [ ] Please put this work behind the feature toggle: NONE
- [ ] This feature should be accessible to the following user groups: ALL
- [ ] Include screenshot(s) in the Github issue if there are front-end changes
- [ ] Update documentation:
- [ ] As a user, I can find all appeal streams by searching for the original appeal's docket number

### Testing Plan
1. Find an appeal whose `stream_docket_number` references something other than itself. With the default DB seed, `Appeal.last.stream_docket_number` was such a case, with its id (1037) being different from the referenced one (201105-1035, your date prefix will probably vary).
2. Without these changes applied, searching (at /search/ in Caseflow) for that should return only one result.
3. With the changes applied, it should find multiple results.
4. Searching on other fields should not change in behavior.

### User Facing Changes
 - [ ] Screenshots of UI changes added to PR & Original Issue

We might be able to just describe this, but:

#### Before

Screen Shot 2020-11-17 at 5 04 54 PM

#### After

Screen Shot 2020-11-17 at 5 05 49 PM

### CODE REVIEW

File: app/services/appeal_finder.rb [hunk 1]
```diff
@@ -18,7 +18,7 @@ def find_appeals_with_file_numbers(file_numbers)
       end
     end
 
-    def find_appeal_by_docket_number(docket_number)
+    def find_appeals_by_docket_number(docket_number)
       return [] unless docket_number
```
Comment (n1zyy) [inline app/services/appeal_finder.rb:22]: My change introduces a small behavior change: if a `docket_number` value is present and not egregiously malformed, but is not in the database, we return an empty ActiveRecord relation. Previously, `find_by` would have raised an exception and then been caught to return nil.

I started to add a test case that, when a docket number was searched for but not found, we should return nil. But then I realized that we return an empty array here if there's no docket number. And CaseSearchResultsForDocketNumber was wrapping the result in an array.

So I'm left unsure if it matters whether we return an empty relationship or nil in that case. On the surface it looks like it doesn't.

File: app/services/appeal_finder.rb [hunk 2]
```diff
@@ -29,9 +29,10 @@ def find_appeal_by_docket_number(docket_number)
         receipt_date = Date.strptime(parsed[0], "%y%m%d")
 
         id = parsed[1]
-        appeal = Appeal.find_by(id: id, receipt_date: receipt_date)
-
-        appeal
+        Appeal.where(
+          "(receipt_date = ? AND id = ?) OR stream_docket_number = ?",
+          receipt_date, id, docket_number
+        )
```
Comment (n1zyy) [inline app/services/appeal_finder.rb:35]: In my dev database, it seems like we could _just_ search on `stream_docket_number`, because it either references another appeal or itself, but doesn't seem to ever be empty. Would that assumption hold up with production data?

If so, this whole method could become:

```ruby
def find_appeals_by_docket_number(docket_number)
  Appeal.find_by(stream_docket_number: docket_number)
end
```

File: app/workflows/case_search_results_for_docket_number.rb [hunk 3]
```diff
@@ -15,7 +15,7 @@ def claim_reviews
   end
 
   def appeals
-    Array.wrap(AppealFinder.find_appeal_by_docket_number(docket_number))
+    AppealFinder.find_appeals_by_docket_number(docket_number)
```
Comment (n1zyy) [inline app/workflows/case_search_results_for_docket_number.rb:18]: Oh, I should change the nil return to `[]` for this to be completely equal.

File: app/services/appeal_finder.rb [hunk 4]
```diff
@@ -29,11 +29,12 @@ def find_appeal_by_docket_number(docket_number)
...
       rescue ArgumentError
-        nil
+        []
```
Comment (n1zyy) [inline app/services/appeal_finder.rb:37]: This method returned nil in some cases and an empty array in others. Our only caller was CaseSearchResultsForDocketNumber.appeals which threw the whole thing inside `Array.wrap()` and thus treated them equally. (AFAICT)

Since `.where` will always return something which is Enumerable, let's standardize on an empty array and get rid of the `Array.wrap`.

File: app/services/appeal_finder.rb (hunk 2 above)
Comment (n1zyy) [reply app/services/appeal_finder.rb:35]: This code change (the one in this PR, not what's described in the comment above) does have the effect of returning an ActiveRecord::Relation instead of an array which may contain an Appeal object.

It appears that the calling code doesn't care; AR::Relation behaves like an Enumerable. Just wanted to flag the change in return type just in case.

File: app/services/appeal_finder.rb [hunk 5]
```diff
@@ -18,7 +18,7 @@ def find_appeals_with_file_numbers(file_numbers)
       end
     end
 
-    def find_appeal_by_docket_number(docket_number)
+    def find_appeals_by_docket_number(docket_number)
```
Comment (n1zyy) [inline app/services/appeal_finder.rb:21]: We support returning more than one Appeal now, hence the pluralization.

This method had only one caller, CaseSearchResultsForDocketNumber, outside of tests, so this was a trivial change to make.

File: app/services/appeal_finder.rb (hunk 2 above)
Comment (hschallhorn) [reply app/services/appeal_finder.rb:35]: This assumption does not hold up unfortunately. `stream_docket_number` is a new addition and is NOW retroactively set any time an appeal is saved, but there is a chance we still have appeals users will search without a stream docket number.

https://github.com/department-of-veterans-affairs/caseflow/blob/15de1d3e26cff73a716e308d0e10d6df09d88e9e/app/models/appeal.rb#L562-L566

Comment (hschallhorn) [reply app/services/appeal_finder.rb:35]: I take that back! All appeals in prod have a stream docket number.

```ruby
Appeal.where(stream_docket_number: nil).count
=> 0
```

File: spec/services/appeal_finder_spec.rb [hunk 6]
```diff
@@ -7,31 +7,45 @@
...
       it "returns nil" do
-        expect(subject).to be_nil
+        expect(subject).to eq([])
+      end
+    end
+
+    context "stream docket number is provided" do
+      let(:docket_number) { second_appeal.stream_docket_number }
+
+      it "returns two object" do
```
Comment (hschallhorn) [inline spec/services/appeal_finder_spec.rb:43]:
```suggestion
      it "returns two appeals" do
```

File: app/workflows/case_search_results_for_docket_number.rb (hunk 3 above)
Comment (hschallhorn) [inline app/workflows/case_search_results_for_docket_number.rb:18]: 👨‍🍳 💋 👌

File: spec/services/appeal_finder_spec.rb [hunk 7]
```diff
@@ -7,31 +7,37 @@
...
+      let(:docket_number) { unknown_docket_number }
 
       it "returns nil" do
-        expect(subject).to be_nil
+        expect(subject).to eq([])
       end
     end
 
-    context "docket number cannot be found" do
-      let(:docket_number) { unknown_docket_number }
```
Comment (n1zyy) [inline spec/services/appeal_finder_spec.rb:31]: The above is a super-confusing way of showing the diff. I removed the "docket number is invalid format" test because it's now the exact same case as "docket number cannot be found."

Also lost from previous discussion: we were previously returning an empty array in some cases and nil in others, but then calling ArrayWrap on the result in the only method that called this. So even though I'm changing the type in the test, the net result is actually the same downstream: an Enumerable of zero or more Appeals.

File: spec/services/appeal_finder_spec.rb [hunk 8]
```diff
@@ -7,31 +7,37 @@
...
-      it "returns nil" do
-        expect(subject).to be_nil
+      it "returns two appeals" do
+        expect(subject.count).to eq(2)
+        expect(subject.first.id).to eq(appeal.id)
+        expect(subject.second.id).to eq(second_appeal.id)
+        expect(subject.first.id).to_not eq(subject.second.id)
+        expect(subject.second.stream_docket_number).to eq(subject.first.stream_docket_number)
       end
     end
```
Comment (n1zyy) [inline spec/services/appeal_finder_spec.rb:42]: I kind of feel like this whole `describe` is now mostly just testing that ActiveRecord works. But I figured I'd let someone else weigh in before being all, "Hi, I'm new here and I completely refactored this method, and then deleted all the tests for it, pls merge." 😉

File: app/services/appeal_finder.rb [hunk 9]
```diff
@@ -69,12 +54,4 @@ def find_appeals_for_vso_user(veterans:)
...
     end
   end
-
-  def filter_appeals_for_vso_user(appeals:, veterans:)
-    allowed_appeal_ids = find_appeals_for_vso_user(veterans: veterans).map(&:id)
-
-    appeals.select do |appeal|
-      allowed_appeal_ids.include?(appeal.id)
-    end
-  end
```
Comment (n1zyy) [inline app/services/appeal_finder.rb:79]: This method was untested. After shortening `find_appeal_by_docket_number`, this accounted for a bit over 10% of the file, bringing us below 90% code coverage. In looking at why none of our test cases invoked this method, it turns out that the answer is that literally nothing calls it anywhere.

File: spec/services/appeal_finder_spec.rb (hunk 8 above)
Comment (hschallhorn) [reply spec/services/appeal_finder_spec.rb:42]: It's a good test to have in case any breaking changes are made to that function. If we ever update it to revert back to the old method, this test will break. Confirming the function returns both appeals is a good confirmation of this feature. I vote keep!

File: app/services/appeal_finder.rb [hunk 10]
```diff
@@ -18,23 +18,8 @@ def find_appeals_with_file_numbers(file_numbers)
...
-        id = parsed[1]
-        appeal = Appeal.find_by(id: id, receipt_date: receipt_date)
-
-        appeal
-      rescue ArgumentError
-        nil
-      end
+    def find_appeals_by_docket_number(docket_number)
+      Appeal.where(stream_docket_number: docket_number)
     end
```
[the remark this answers is not in the document]
Comment (hschallhorn) [reply app/services/appeal_finder.rb:23]: Mind adding a comment explaining why this function will return both the original appeal and any appeal streams off of the original?

**Review by hschallhorn — APPROVED** [review]
Screen Shot 2020-11-23 at 4 36 13 PM
 🙌  Nice catch on simplifying the method!
